### DISCUSSION TITLE
[Proposal]: nested members in `with` and object creation

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#9002] [Language Ideas]
# nested members in `with` and object creation

Champion issue: https://github.com/dotnet/csharplang/issues/9003

## Summary
[summary]: #summary

If we go ahead and add support for [extended names in property patterns](https://github.com/dotnet/csharplang/issues/4394) (`if (e is MethodCallExpression { Method.Name: "MethodName" })`), it makes sense for symmetry to allow this in `with` and object creation expressions:

`methodCallExpression with { Method.Name = "MethodName" }`
`new MethodCallExpression(...) { Method.Name = "MethodName" }`

## Motivation
[motivation]: #motivation

This is particularly useful in the `with` expression, so you don't have to write:
`methodCallExpression with { Method = methodCallExpression.Method with { Name = "MethodName" } }`

## Detailed design
[design]: #detailed-design

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

## Design meetings

* https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-09-22.md#nested-members-in-with-and-object-creation
* https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-01-24.md#nested-members-in-with-and-object-creation

**Comment by jnm2** [comment]
Existing discussion on nested `with` expressions: https://github.com/dotnet/csharplang/discussions/77

Could there be names in scope to avoid doing a deep traversal twice? A common scenario (usually deeper):

```cs
// Names in scope after '=': NestedThing, Collection
thing = thing with { NestedThing.Collection = Collection.Remove(value) };
```

As opposed to:
```cs
thing = thing with { NestedThing.Collection = thing.NestedThing.Collection.Remove(value) };
```

**Comment by DustinCampbell** [comment]
Here's another potential alternate syntax:

```C#
methodCallExpression with
{
    Method with { Name = "MethodName" }
}
```

**Comment by XDracam** [comment]
I hacked together my own Lenses with source generators today just to avoid annoying nested `with` statements. If anyone needs a quick workaround, [here it is](https://github.com/XDracam/FastOptics) (MIT license, use at your own risk). But I'd prefer a convenient builtin syntax.
